File: src/linkage_sim/analysis/crank_selection.py
"""Crank selection analysis: recommend which link to drive for maximum rotation.

For 4-bar mechanisms, uses Grashof classification to determine the optimal
crank analytically. For general mechanisms (6-bar, etc.), uses numerical
probing to estimate the valid angular range of each candidate driven link.

The user can always override the recommendation by specifying their own
revolute driver -- these functions are advisory, not prescriptive.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np
from numpy.typing import NDArray

from linkage_sim.analysis.grashof import GrashofType, check_grashof
from linkage_sim.core.constraints import RevoluteJoint
from linkage_sim.core.drivers import RevoluteDriver
from linkage_sim.core.mechanism import Mechanism
from linkage_sim.core.state import GROUND_ID


@dataclass(frozen=True)
class FourbarTopology:
    """Detected topology parameters for a 4-bar mechanism.

    Attributes:
        ground_length: Distance between ground pivots.
        ground_adjacent: Map of body_id -> link_length for ground-adjacent bodies.
        coupler_id: ID of the coupler body (not connected to ground).
        coupler_length: Length of the coupler link.
        driven_body_id: ID of the driven body (from RevoluteDriver), or None.
    """

    ground_length: float
    ground_adjacent: dict[str, float]
    coupler_id: str
    coupler_length: float
    driven_body_id: str | None


@dataclass(frozen=True)
class CrankRecommendation:
    """Recommendation for which link to drive in a mechanism.

    Attributes:
        body_id: ID of the recommended crank body.
        estimated_range_deg: Estimated angular range in degrees (360 = full rotation).
        full_rotation: True if the link can complete a full revolution.
        reason: Human-readable explanation of why this recommendation was made.
    """

    body_id: str
    estimated_range_deg: float
    full_rotation: bool
    reason: str


def detect_fourbar_topology(
    mechanism: Mechanism,
) -> FourbarTopology | None:
    """Detect whether a mechanism has 4-bar topology and extract its parameters.

    A valid 4-bar has:
    - Exactly 3 moving bodies (plus ground)
    - Exactly 4 RevoluteJoint instances (not counting RevoluteDriver)
    - Exactly 2 of those joints connect to ground

    Args:
        mechanism: A Mechanism instance (built or unbuilt, with or without driver).

    Returns:
        A FourbarTopology with topology info if the mechanism is a 4-bar,
        or None otherwise.
    """
    bodies = mechanism.bodies
    moving_body_ids = [bid for bid in bodies if bid != GROUND_ID]

    # Must have exactly 3 moving bodies
    if len(moving_body_ids) != 3:
        return None

    # Count RevoluteJoint instances (not RevoluteDriver)
    revolute_joints: list[RevoluteJoint] = [
        j for j in mechanism.joints if isinstance(j, RevoluteJoint)
    ]
    if len(revolute_joints) != 4:
        return None

    # Find joints that connect to ground
    ground_joints: list[RevoluteJoint] = [
        j
        for j in revolute_joints
        if j.body_i_id == GROUND_ID or j.body_j_id == GROUND_ID
    ]
    if len(ground_joints) != 2:
        return None

    # Identify ground-adjacent bodies (connected to ground via revolute joint)
    ground_adjacent: dict[str, float] = {}
    ground_attachment_points: list[str] = []

    for gj in ground_joints:
        if gj.body_i_id == GROUND_ID:
            adjacent_body_id = gj.body_j_id
            ground_pt_name = gj._point_i_name
        else:
            adjacent_body_id = gj.body_i_id
            ground_pt_name = gj._point_j_name

        # Compute link length from attachment points
        body = bodies[adjacent_body_id]
        pts = list(body.attachment_points.values())
        link_length = float(np.linalg.norm(pts[1] - pts[0]))
        ground_adjacent[adjacent_body_id] = link_length
        ground_attachment_points.append(ground_pt_name)

    # Compute ground length from first two ground attachment points
    ground_body = bodies[GROUND_ID]
    ground_pts = list(ground_body.attachment_points.values())
    ground_length = float(np.linalg.norm(ground_pts[1] - ground_pts[0]))

    # Identify coupler (the moving body not connected to ground)
    coupler_ids = [bid for bid in moving_body_ids if bid not in ground_adjacent]
    if len(coupler_ids) != 1:
        return None
    coupler_id = coupler_ids[0]

    coupler_body = bodies[coupler_id]
    coupler_pts = list(coupler_body.attachment_points.values())
    coupler_length = float(np.linalg.norm(coupler_pts[1] - coupler_pts[0]))

    # Detect driven body from RevoluteDriver
    driven_body_id: str | None = None
    for j in mechanism.joints:
        if isinstance(j, RevoluteDriver):
            # The driven body is the non-ground body in the driver
            if j.body_i_id == GROUND_ID:
                driven_body_id = j.body_j_id
            elif j.body_j_id == GROUND_ID:
                driven_body_id = j.body_i_id
            else:
                # Driver between two moving bodies; pick body_j as driven
                driven_body_id = j.body_j_id
            break

    return FourbarTopology(
        ground_length=ground_length,
        ground_adjacent=ground_adjacent,
        coupler_id=coupler_id,
        coupler_length=coupler_length,
        driven_body_id=driven_body_id,
    )


def _estimate_driven_range_analytical(
    ground_len: float,
    crank_len: float,
    coupler_len: float,
    rocker_len: float,
) -> float:
    """Estimate the angular range of the crank that produces valid assembly.

    Scans 3600 evenly spaced crank angles and counts how many produce a
    valid diagonal distance d_tip that satisfies the triangle inequality
    with the coupler and rocker.

    Args:
        ground_len: Distance between ground pivots.
        crank_len: Length of the candidate crank link.
        coupler_len: Length of the coupler link.
        rocker_len: Length of the rocker link.

    Returns:
        Estimated angular range in degrees (0 to 360).
    """
    n_samples = 3600
    thetas = np.linspace(0.0, 2.0 * np.pi, n_samples, endpoint=False)

    # Distance from ground pivot to crank tip
    d_tip = np.sqrt(
        crank_len**2
        + ground_len**2
        - 2.0 * crank_len * ground_len * np.cos(thetas)
    )

    # Triangle inequality for coupler + rocker to close the loop
    lower = np.abs(coupler_len - rocker_len)
    upper = coupler_len + rocker_len

    valid_count = int(np.sum((d_tip >= lower) & (d_tip <= upper)))
    return (valid_count / n_samples) * 360.0


def recommend_crank_fourbar(mechanism: Mechanism) -> list[CrankRecommendation]:
    """Recommend which ground-adjacent link to drive for maximum rotation.

    Uses Grashof classification to determine the optimal crank for 4-bar
    mechanisms. Returns recommendations ranked best-first (full rotation
    candidates first, then by range descending).

    Args:
        mechanism: A Mechanism instance (built or unbuilt).

    Returns:
        List of CrankRecommendation sorted best-first.
        Empty list if the mechanism is not a 4-bar.
    """
    topo = detect_fourbar_topology(mechanism)
    if topo is None:
        return []

    ground_length = topo.ground_length
    ground_adjacent = topo.ground_adjacent
    coupler_length = topo.coupler_length

    recommendations: list[CrankRecommendation] = []

    for body_id, link_length in ground_adjacent.items():
        # Determine which is "crank" and which is "rocker" for this trial
        other_ids = [bid for bid in ground_adjacent if bid != body_id]
        if len(other_ids) != 1:
            continue
        other_length = ground_adjacent[other_ids[0]]

        # Call check_grashof with this body as the "crank" position
        result = check_grashof(
            ground_length=ground_length,
            crank_length=link_length,
            coupler_length=coupler_length,
            rocker_length=other_length,
        )

        if result.classification == GrashofType.GRASHOF_CRANK_ROCKER:
            # Full rotation only if shortest link is the crank
            if result.shortest_is == "crank":
                recommendations.append(
                    CrankRecommendation(
                        body_id=body_id,
                        estimated_range_deg=360.0,
                        full_rotation=True,
                        reason=(
                            f"Grashof crank-rocker: '{body_id}' is the shortest "
                            f"link ({link_length}) and can fully rotate."
                        ),
                    )
                )
            else:
                # Crank-rocker but this link is the rocker side
                est_range = _estimate_driven_range_analytical(
                    ground_length, link_length, coupler_length, other_length
                )
                recommendations.append(
                    CrankRecommendation(
                        body_id=body_id,
                        estimated_range_deg=est_range,
                        full_rotation=False,
                        reason=(
                            f"Grashof crank-rocker: '{body_id}' is not the "
                            f"shortest link; limited to ~{est_range:.1f} deg."
                        ),
                    )
                )
        elif result.classification == GrashofType.GRASHOF_DOUBLE_CRANK:
            recommendations.append(
                CrankRecommendation(
                    body_id=body_id,
                    estimated_range_deg=360.0,
                    full_rotation=True,
                    reason=(
                        f"Grashof double-crank: ground is shortest link, "
                        f"'{body_id}' can fully rotate."
                    ),
                )
            )
        elif result.classification == GrashofType.CHANGE_POINT:
            recommendations.append(
                CrankRecommendation(
                    body_id=body_id,
                    estimated_range_deg=360.0,
                    full_rotation=True,
                    reason=(
                        f"Change-point: '{body_id}' can fully rotate "
                        f"(may lock at dead points)."
                    ),
                )
            )
        else:
            # GRASHOF_DOUBLE_ROCKER or NON_GRASHOF: limited range
            est_range = _estimate_driven_range_analytical(
                ground_length, link_length, coupler_length, other_length
            )
            recommendations.append(
                CrankRecommendation(
                    body_id=body_id,
                    estimated_range_deg=est_range,
                    full_rotation=False,
                    reason=(
                        f"'{body_id}' cannot fully rotate; "
                        f"estimated range ~{est_range:.1f} deg."
                    ),
                )
            )

    # Sort: full_rotation first, then by range descending
    recommendations.sort(
        key=lambda r: (not r.full_rotation, -r.estimated_range_deg)
    )

    return recommendations


def estimate_driven_range(
    mechanism: Mechanism,
    q0: NDArray[np.float64],
    n_probes: int = 72,
) -> float:
    """Estimate valid angular range of the current driver by probing.

    Attempts to solve the mechanism at n_probes evenly spaced angles
    (0 to 360 degrees). Returns the estimated range in degrees based
    on how many probes converge.

    This works for any mechanism topology (4-bar, 6-bar, etc.).

    Assumes the driver uses f(t) = t (identity), so t maps directly
    to the driven angle in radians. This is the convention used by
    all viewer scripts.

    Args:
        mechanism: A built Mechanism with an identity revolute driver.
        q0: Initial guess for the solver.
        n_probes: Number of angles to test (default 72 = every 5 degrees).

    Returns:
        Estimated valid range in degrees (0 to 360).
    """
    from linkage_sim.solvers.kinematics import solve_position

    angles = np.linspace(0, 2 * np.pi, n_probes, endpoint=False)
    n_converged = 0
    q_current = q0.copy()

    for angle in angles:
        result = solve_position(mechanism, q_current, t=float(angle))
        if result.converged:
            n_converged += 1
            q_current = result.q.copy()

    return n_converged / n_probes * 360.0

File: linkage-sim-rs/src/analysis/crank_selection.rs
//! Crank selection analysis: recommend which link to drive for maximum rotation.
//!
//! For 4-bar mechanisms, uses Grashof classification to determine the optimal
//! crank analytically. Each ground-adjacent link is evaluated as a candidate
//! driver, and candidates are ranked by rotation capability.

use super::grashof::{check_grashof, GrashofType};

/// A candidate link that could serve as the driven crank.
#[derive(Debug, Clone)]
pub struct CrankCandidate {
    /// Label of this link (e.g. "ground", "crank", "coupler", "rocker").
    pub link_name: String,
    /// Length of this link.
    pub link_length: f64,
    /// Whether this link can fully rotate (360 degrees).
    pub can_fully_rotate: bool,
    /// Whether this link is the shortest in the mechanism.
    pub is_shortest: bool,
    /// Estimated angular range in degrees (360.0 for full rotation).
    pub estimated_range_deg: f64,
    /// Human-readable explanation of this candidate's capability.
    pub reason: String,
}

/// Recommendation for which link to use as the crank (driver).
#[derive(Debug, Clone)]
pub struct CrankRecommendation {
    /// Recommended driver link name.
    pub recommended_link: String,
    /// Whether the recommended link can rotate fully (360 degrees).
    pub full_rotation: bool,
    /// Grashof classification of the mechanism.
    pub classification: GrashofType,
    /// All candidates with their rotation capability, sorted best-first.
    pub candidates: Vec<CrankCandidate>,
}

/// Estimate the angular range of a candidate crank by numerical probing.
///
/// Scans evenly-spaced crank angles and checks whether the diagonal distance
/// between the crank tip and the rocker pivot satisfies the triangle inequality
/// with the coupler and rocker. Returns estimated range in degrees (0 to 360).
fn estimate_driven_range(
    ground_len: f64,
    crank_len: f64,
    coupler_len: f64,
    rocker_len: f64,
) -> f64 {
    let n_samples: usize = 3600;
    let lower = (coupler_len - rocker_len).abs();
    let upper = coupler_len + rocker_len;

    let mut valid_count: usize = 0;
    for i in 0..n_samples {
        let theta = (i as f64) * 2.0 * std::f64::consts::PI / (n_samples as f64);
        let d_tip = (crank_len * crank_len + ground_len * ground_len
            - 2.0 * crank_len * ground_len * theta.cos())
        .sqrt();
        if d_tip >= lower && d_tip <= upper {
            valid_count += 1;
        }
    }

    (valid_count as f64 / n_samples as f64) * 360.0
}

/// For a 4-bar mechanism, recommend which link to drive.
///
/// Evaluates each ground-adjacent link (crank and rocker) as a potential driver
/// using Grashof's theorem. The shortest link can fully rotate if and only if
/// S + L <= P + Q (Grashof condition).
///
/// Rules:
/// - CrankRocker: drive the shortest link (it's the crank that fully rotates)
/// - DoubleCrank: either grounded link works; both fully rotate
/// - DoubleRocker: no link can fully rotate; pick the one with greater range
/// - ChangePoint: any configuration possible but may have dead points
/// - NonGrashof: no full rotation; pick the link with greater range
///
/// The four lengths are given in the conventional order: ground (fixed frame
/// distance between pivots), crank (input), coupler (floating), rocker (output).
pub fn recommend_crank(
    ground_length: f64,
    crank_length: f64,
    coupler_length: f64,
    rocker_length: f64,
) -> CrankRecommendation {
    let result = check_grashof(
        ground_length,
        crank_length,
        coupler_length,
        rocker_length,
        1e-10,
    );

    let links: [(&str, f64); 4] = [
        ("ground", ground_length),
        ("crank", crank_length),
        ("coupler", coupler_length),
        ("rocker", rocker_length),
    ];

    let shortest_len = links
        .iter()
        .map(|(_, l)| *l)
        .fold(f64::INFINITY, f64::min);

    // Build candidates for ground-adjacent links only (crank and rocker).
    // Ground and coupler cannot serve as drivers in a standard 4-bar.
    let mut candidates: Vec<CrankCandidate> = Vec::new();

    for &(name, len) in &[("crank", crank_length), ("rocker", rocker_length)] {
        let is_shortest = (len - shortest_len).abs() < 1e-10;

        // When this link is used as the driver, the "other" grounded link
        // becomes the output. Compute Grashof classification with this
        // link in the crank position.
        let other_len = if name == "crank" {
            rocker_length
        } else {
            crank_length
        };

        let trial =
            check_grashof(ground_length, len, coupler_length, other_len, 1e-10);

        let (can_fully_rotate, estimated_range_deg, reason) = match trial.classification {
            GrashofType::CrankRocker => {
                if trial.shortest_is == "crank" {
                    // This link is shortest and can fully rotate.
                    (
                        true,
                        360.0,
                        format!(
                            "Crank-rocker: '{}' is the shortest link ({:.3}) and can fully rotate.",
                            name, len
                        ),
                    )
                } else {
                    let range = estimate_driven_range(
                        ground_length,
                        len,
                        coupler_length,
                        other_len,
                    );
                    (
                        false,
                        range,
                        format!(
                            "Crank-rocker: '{}' is not the shortest link; limited to ~{:.1} deg.",
                            name, range
                        ),
                    )
                }
            }
            GrashofType::DoubleCrank => (
                true,
                360.0,
                format!(
                    "Double-crank: ground is shortest, '{}' can fully rotate.",
                    name
                ),
            ),
            GrashofType::ChangePoint => (
                true,
                360.0,
                format!(
                    "Change-point: '{}' can fully rotate (may lock at dead points).",
                    name
                ),
            ),
            _ => {
                // DoubleRocker or NonGrashof: limited range.
                let range = estimate_driven_range(
                    ground_length,
                    len,
                    coupler_length,
                    other_len,
                );
                (
                    false,
                    range,
                    format!(
                        "'{}' cannot fully rotate; estimated range ~{:.1} deg.",
                        name, range
                    ),
                )
            }
        };

        candidates.push(CrankCandidate {
            link_name: name.to_string(),
            link_length: len,
            can_fully_rotate,
            is_shortest,
            estimated_range_deg,
            reason,
        });
    }

    // Sort: full-rotation candidates first, then by range descending.
    candidates.sort_by(|a, b| {
        let key_a = (!a.can_fully_rotate, -a.estimated_range_deg as i64);
        let key_b = (!b.can_fully_rotate, -b.estimated_range_deg as i64);
        key_a.cmp(&key_b)
    });

    let recommended = candidates
        .first()
        .map(|c| c.link_name.clone())
        .unwrap_or_else(|| "crank".to_string());
    let full_rotation = candidates
        .first()
        .is_some_and(|c| c.can_fully_rotate);

    CrankRecommendation {
        recommended_link: recommended,
        full_rotation,
        classification: result.classification,
        candidates,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn crank_rocker_recommends_shortest_link() {
        // ground=4, crank=1, coupler=3.5, rocker=2.5
        // S=1 (crank), L=4, P=2.5, Q=3.5 => S+L=5, P+Q=6 => Grashof, crank-rocker
        let rec = recommend_crank(4.0, 1.0, 3.5, 2.5);
        assert_eq!(rec.classification, GrashofType::CrankRocker);
        assert_eq!(rec.recommended_link, "crank");
        assert!(rec.full_rotation);
        assert_eq!(rec.candidates.len(), 2);

        // The recommended candidate (crank) should have full rotation.
        let best = &rec.candidates[0];
        assert!(best.can_fully_rotate);
        assert!((best.estimated_range_deg - 360.0).abs() < 1e-6);

        // The other candidate (rocker) should have limited range.
        let other = &rec.candidates[1];
        assert!(!other.can_fully_rotate);
        assert!(other.estimated_range_deg < 360.0);
        assert!(other.estimated_range_deg > 0.0);
    }

    #[test]
    fn double_crank_both_links_can_fully_rotate() {
        // ground=1, crank=2.5, coupler=3.5, rocker=4
        // S=1 (ground), L=4, P=2.5, Q=3.5 => S+L=5, P+Q=6 => Grashof, double-crank
        let rec = recommend_crank(1.0, 2.5, 3.5, 4.0);
        assert_eq!(rec.classification, GrashofType::DoubleCrank);
        assert!(rec.full_rotation);

        // Both ground-adjacent links should fully rotate.
        for c in &rec.candidates {
            assert!(
                c.can_fully_rotate,
                "In double-crank, '{}' should fully rotate",
                c.link_name
            );
            assert!(
                (c.estimated_range_deg - 360.0).abs() < 1e-6,
                "Expected 360 deg for '{}', got {}",
                c.link_name,
                c.estimated_range_deg
            );
        }
    }

    #[test]
    fn non_grashof_no_full_rotation() {
        // ground=1, crank=5, coupler=2, rocker=2
        // S=1, L=5, P=2, Q=2 => S+L=6, P+Q=4 => NOT Grashof
        let rec = recommend_crank(1.0, 5.0, 2.0, 2.0);
        assert_eq!(rec.classification, GrashofType::NonGrashof);
        assert!(!rec.full_rotation);

        // No candidate should have full rotation.
        for c in &rec.candidates {
            assert!(
                !c.can_fully_rotate,
                "Non-Grashof '{}' should not fully rotate",
                c.link_name
            );
        }
    }

    #[test]
    fn change_point_allows_full_rotation() {
        // ground=4, crank=1, coupler=3, rocker=2
        // S=1, L=4, P=2, Q=3 => S+L=5, P+Q=5 => change-point
        let rec = recommend_crank(4.0, 1.0, 3.0, 2.0);
        assert_eq!(rec.classification, GrashofType::ChangePoint);
        assert!(rec.full_rotation);

        // At least one candidate should have full rotation.
        let any_full = rec.candidates.iter().any(|c| c.can_fully_rotate);
        assert!(any_full, "Change-point should have at least one full-rotation candidate");
    }
}
